tools/nl/embeddings/build_embeddings_v2_test.py

```diff
@@ -137,3 +158,126 @@ def testSuccess(self):
+    self.assertEqual(len(dcid_sentence_df),
+                     len(dcid_sentence_df["dcid"].unique()))
+    self.assertEqual(len(dcid_sentence_df),
+                     len(dcid_sentence_df["sentence"].unique()))
+
+    # Check that there are no duplicate sentences.
+    sentences = {}
+    for row_index, row in dcid_sentence_df.iterrows():
+      sv_dcid = row["dcid"]
+      for s in row["sentence"].split(";"):
+        if not s:
+          continue
+
+        existing_sv_index = sentences.get(s, ())
+        err_msg = f"\nSentence already exists: {s}.\nCurrent (SV, index): {sv_dcid, row_index}.\nExisting (SV, index): {existing_sv_index}."
+        self.assertFalse(s in sentences, err_msg)
+        sentences[s] = (sv_dcid, row_index)
+
+  def testEmbeddingsValidations(self):
+    # TODO: consider moving this test to a separate file which is
+    # tiggered on demand. This test can take a while to run (esp when the
+    # number of SVs and sentences becomes larger in the future.
+    sheets_filepath = os.path.join("sheets", "dc_nl_svs_curated.csv")
+    input_other_alternatives_filepath = os.path.join("csv",
+                                                     "other_alternatives.csv")
+    input_palm_alternatives_filepath = os.path.join("csv",
+                                                    "palm_alternatives.csv")
+    output_dcid_sentences_filepath = os.path.join(
+        "csv", "embeddings_input_sv_descriptions.csv")
+
+    # Mock the get_sheets_data function to return data from sheets_filepath.
+    be.get_sheets_data = mock.Mock(
+        return_value=pd.read_csv(sheets_filepath).fillna(""))
+
+    # Given that the get_sheets_data() function is mocked, the Context
+    # object does not need a valid `gs` and `bucket` field.
+    ctx = be.Context(gs=None,
+                     model=SentenceTransformer("all-MiniLM-L6-v2"),
+                     bucket="",
+                     tmp="/tmp")
+
+    # Smilarly, sheets_url and worksheet_name can be empty strings.
+    sheets_url = ""
+    worksheet_name = ""
+
+    # Intermediate outputs can be written to a temporary directory.
+    with tempfile.TemporaryDirectory() as tmp_dir:
+      tmp_local_sheets_csv_filepath = os.path.join(tmp_dir, "sheets_data.csv")
+      tmp_local_merged_filepath = os.path.join(tmp_dir, "merged_data.csv")
+
+      embeddings_df = be.build(
+          ctx, sheets_url, worksheet_name, tmp_local_sheets_csv_filepath,
+          tmp_local_merged_filepath,
+          [input_other_alternatives_filepath, input_palm_alternatives_filepath])
+
+      # Verify that embeddings were created for all DCIDs and Sentences.
+      dcid_sentence_df = pd.read_csv(output_dcid_sentences_filepath).fillna("")
+      sentences = set()
+      for alts in dcid_sentence_df["sentence"].values:
+        for s in alts.split(";"):
+          if not s:
+            continue
+          sentences.add(s)
+
+      self.assertEqual(len(sentences), len(embeddings_df))
```

Comment:
If this fails, wonder how we would go about debugging?  Instead, how about actually comparing the list of sentences, appropriately ordered?  (so the output will emit the delta)

Reply:
Good point. Updated the test